File: 02/2.py
def main():
    n = 1000
    a = []
    for i in range(1000):
        a.append(list(map(int, input().split())))

    ans = 0
    for line in a:
        for i in range(len(line)):
            l = line[:i] + line[i + 1 :]
            f = 1
            if l[1] > l[0]:
                for x, y in zip(l, l[1:]):
                    if y - x > 3 or y - x < 1:
                        f = 0

            else:
                for x, y in zip(l[1:], l):
                    if y - x > 3 or y - x < 1:
                        f = 0
            if f == 1:
                ans += 1
                break

    print(ans)


if __name__ == "__main__":
    main()

File: 19/2.hs
import System.IO
import Data.List (isPrefixOf)
import Data.List.Split
import Debug.Trace (trace)
import qualified Data.Map as Map
import Data.Map (Map)

main::IO ()
main = do
  let  filename = "1.in"
  content <- readFile filename
  let ans = count content
  print  ans


parse :: String -> ([String], [String])
parse input = (tokens , liness)
  where
  [part1 , part2] = splitOn "\n\n" input
  tokens = splitOn ", " part1
  liness = lines part2

check :: String -> [String] -> Map.Map String Int -> (Int , Map.Map String Int)
check "" _ cache = (1, cache)

check line tokens cache =
    case Map.lookup line cache of
        Just result -> (result, cache)
        Nothing ->
            let
                (result, newCache) = foldl checkToken (0, cache) tokens
                checkToken (count, cache') token
                    | token `isPrefixOf` line =
                        let (res, updatedCache) = check (drop (length token) line) tokens cache'
                        in (count + res, updatedCache)
                    | otherwise = (count, cache')
                updatedCache = Map.insert line result newCache
            in (result, updatedCache)


checkWithMemo :: String -> [String] -> Int
checkWithMemo line tokens = fst $ check line tokens Map.empty

count :: String -> Int
count input = sum $ map (`checkWithMemo` tokens) liness
  where
    (tokens, liness) = parse input
